enhancements/image-references.md

```diff
@@ -0,0 +1,228 @@
+repository that are self-consistent, valid and make no assumptions about the
+build system that will eventually create a bundle image with them.
+
+#### Story 2
+
+As an author of an operand, I would like to create a bundle locally in order
+to test my operator end-to-end on a cluster of my choosing without having to
+edit the core configuration for the operator.
+
+#### Story 3
+
+As an author of a build system, I would like to operate with tooling that allows
+me to clearly define the source of truth for image digests in order to keep the
+build-system-specific configuration to an absolute minimum.
+
+#### Story 4
+
+As an engineer involved in publishing an optional operator, I would like to
+configure semantically equivalent image pull specifications once, in order to not
+need to configure each build system independently.
+
+### Implementation Details/Notes/Constraints
+
+The core problems that must be solved in the implementation of this proposal have
+already been handled in the workflow used in `oc adm release new`. When implementing
+improvements to the `operator-sdk bundle create` process we will simply need to
+create a shared library for the two tools to use. While the shape of the output
+is slightly different and some of the semantics about how the output should be 
+formatted are dissmilar, the core image reference rewriting is identical and the
+process of building a `FROM scratch` image layer is also identical.
+
+Today, some prior art exists in the OSBS workflow for building operator bundles.
+As we improve the Operator SDK tooling to create a straghtforward process for
+creating bundle images, we must make sure a seamless migration is possible.
+
+### Risks and Mitigations
+
+It will be critical that the design be vetted by all of the concerned parties, from
+operator manifest authors to CI system authors and productized pipeline authors to
+ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
+implementation here is not prioritizing a clean migration pathway for all current
+users who create bundle images, which would lead to further fragmentation of the
+ecosystem, which is directly opposed to the goal of this enhancement.
+
+## Design Details
+
+The definition of a minimally-viable operator bundle image will be changed to
+ensure that all iamge references in the contained manifests have been resolved 
+to a digest and had the pull specifications rewritten to refer to those digests.
+
+The only acceptable process for creating an operator bundle will be to run the
+`operator-sdk bundle create` CLI, providing the manifests, metadata and image
+sources as input to the creation process.
+
+### Proposed UX
+
+Operator manifest authors write a manifest that refers to images using some
+opaque string [ex](https://github.com/openshift/machine-config-operator/blob/master/install/0000_80_machine-config-operator_04_deployment.yaml#L22),
```

Comment:
Can we be clear about the wording here and say `image tag`? Saying `opaque string`, while general, implies that you can set this to anything which could break a Deployment that is used both in and outside of a bundle.

Reply:
This is explicitly not an image tag. Why should we care to support the Deployment outside of a Bundle?

Reply:
I have an operator project that I would like to bundle. I modify my Deployment's operator image tag to be an opaque string. Now I can bundle my operator successfully, but running `kubectl apply -f deployment.yaml` breaks unless I keep 2 versions of my Deployment or modify it after generating a CSV. We should not be breaking existing operator projects for image references.

Instead, I should be able to leave my image tag as-is, which is effectively:
```yaml
imageReferences:
- name: registry.svc.ci.openshift.org/openshift:image
  placeholder: registry.svc.ci.openshift.org/openshift:image
```

So we can simplify the image references spec to:
```yaml
imageReferences:
- registry.svc.ci.openshift.org/openshift:image
```
because wouldn't the author want all references to the same tag be updated in the same way? This string is just as easy to reference in `image-replacements`.

Reply:
Why do we care to support `kubectl apply`? The entire point of the operator is to deploy all of the manifests together. Why should a developer want to apply some small subset of their manifests instead of running a (trivially short) bundle creation and deploying that way, even for their development work?

Reply:
I explicitly do not want to see us guessing at the common names here. Allowing an explicit mapping between a common name and the opaque string is required to interoperate between separate build systems which may call their images by different names, but need a common name as currency to uniquely identify the container image regardless.

Reply:
Furthermore -- the opqaue string may happen to be a valid pull spec, so if a packager were to use an upstream set of manifests verbatim, there would be no backwards-incompatible change from this proposal and no breaking for existing operator projects.

Reply:
`kubectl apply` was just an example. Given a new operator project, I should be able to run
```sh
$ kustomize build config/default | kubectl apply -f -
```
which includes all operator manifests required to run my operator.

The explicit mapping is fine, I'm mainly arguing against saying this is an opaque string because it can be a valid pull spec which is not opaque. Given your last point, `placeholder` could be optional in case you're using a valid pull spec. Then the first example above would look like:
```yaml
imageReferences:
- name: registry.svc.ci.openshift.org/openshift:image
```

Maybe we're just arguing over syntax here.

Reply:
> Given a new operator project, I should be able to run

Why?

Reply:
This is what [`make deploy`](https://github.com/operator-framework/operator-sdk-samples/blob/master/go/kubebuilder/memcached-operator/Makefile#L46) does in the [new project layout](https://master.sdk.operatorframework.io/docs/golang/quickstart/#build-and-run-the-operator). `operator-sdk` needs to be a drop-in replacement for `kubebuilder` while also supporting bundles, so the ability to do this can/should not change.

Reply:
Makes sense. Keeping backwards compatibility here should be simple, and I agree we should strive to do so.